### DISCUSSION TITLE
State observer

### DISCUSSION THREAD

**Discussion by ValentinKaisermayer** [ReactiveBayes/RxInfer.jl#309] [Q&A] [answered]
I have a simple spring-mass-damper system for which I want to make a state-observer:
```julia
using RxInfer
using Random
using LinearAlgebra
using CairoMakie

function f(x, u, p, t, tS)
    return x + tS*[
        x[2]
        -p.c / p.m * x[1] - p.d / p.m * x[2] - 1 / p.m * u[1]
    ]
end
g(x, u, p, t) = [x[1]]

tS = 1/100
t_sim = 0:tS:120
u_sim = reduce(
    hcat,
    map(t -> [exp(-0.1t) * sin(1 / 10 * 2 * pi * t) + sin(1 / 20 * 2 * pi * t)], t_sim),
)
x0 = [0, 0]

function generate_data(rng, f, g, tS, x0, u, t, Q, R)
    x_prev = x0

    x = Vector{Vector{Float64}}(undef, length(t))
    y = Vector{Vector{Float64}}(undef, length(t))

    for i in 1:length(t)
        x[i] = rand(rng, MvNormal(f(x_prev, u[:,i], p, t[i], tS), Q))
        y[i] = rand(rng, MvNormal(g(x[i], u[:,i], p, t[i]), R))
        x_prev = x[i]
    end
    
    return x, y
end

seed = 1234
rng = MersenneTwister(1234)

Q = 0.1 .* diageye(2)
R = 10 .* diageye(1)
p = (c=1.0, d=0.1, m=0.8)

x, y = generate_data(rng, f, g, tS, x0, u_sim, t_sim, Q, R)

let 
    fig = Figure()
    ax1 = Axis(fig[1,1])
    lines!(t_sim, getindex.(x, 1), label = "Hidden Signal (dim-1)", color = :teal)
    lines!(t_sim, getindex.(x, 2), label = "Hidden Signal (dim-2)", color = :violet)
    ax2 = Axis(fig[2,1])
    lines!(t_sim, getindex.(y, 1), label = "Observation", color = :orange)

    fig
end
```

![fig](https://github.com/ReactiveBayes/RxInfer.jl/assets/50108075/8e3871c8-5d71-4d5a-802c-234a5d388d43)

However, I have some problems with the model specification:

```julia
@model function ssm(y, f, g, tS, x0, u, t, Q, R)
    x_prior ~ x0

    x_prev = x_prior
    for i in eachindex(t)
        x[i] ~ MvNormalMeanCovariance(mu=f(x_prev, u[:,i], p, t[i], tS), var=Q)
        y[i] ~ MvNormalMeanCovariance(mu=g(x[i], u[:,i], p, t[i]), var=R)
        x_prev = x[i]
    end
end

# prior_x: Based on first observation, assuming initial state is similar with equal variance.
prior_x = MvNormalMeanCovariance(x[1], diageye(2))

model = ssm(f=f, g=g, tS=tS, x0=prior_x, u=u_sim, t=t_sim, Q=Q, R=R)

ssm_meta = @meta begin 
    f() -> Unscented()
    g() -> Unscented()
end

imessages = @initialization begin
    μ(x) = prior_x
end

result = infer(
    model = model,
    data  = (y=y, ), 
    options = (limit_stack_depth = 500, ), 
    returnvars = KeepLast(),
    predictvars = KeepLast(),
    initialization = imessages,
    meta = ssm_meta,
    iterations = 20,
    showprogress=true,
)
```

```julia
ERROR: AssertionError: Expected only one missing interface, got (:out, :μ, :Σ) of length 3 (node MvNormal with interfaces (:mu, :var))
Stacktrace:
  [1] prepare_interfaces(::GraphPPL.StaticInterfaces{…}, fform::Type{…}, lhs_interface::GraphPPL.NodeLabel, rhs_interfaces::@NamedTuple{…})
    @ GraphPPL C:\Users\KaisermayerV\.julia\packages\GraphPPL\Z49xA\src\graph_engine.jl:1622
```

Eventually, I would like to infer an unknown `p.m` from given data.

**Comment by ValentinKaisermayer** [comment]
It should be:
```julia
@model function ssm(y, f, g, tS, x0, u, t, Q, R)
    x_prior ~ x0

    x_prev = x_prior
    for i in eachindex(t)
        x[i] ~ MvNormalMeanCovariance(μ=f(x_prev, u[:,i], p, t[i], tS), Σ=Q)
        y[i] ~ MvNormalMeanCovariance(μ=g(x[i], u[:,i], p, t[i]), Σ=R)
        x_prev = x[i]
    end
end
```
However, it does not terminate.

**Comment by ValentinKaisermayer** [comment]
I have something working; however, the results are a bit underwhelming, considering that this is just the same as the linear Kalman filter example.
```julia
result = infer(
    model = model,
    data  = (y=y, ), 
    options = (limit_stack_depth = 10, ), 
    initialization = imessages,
    meta = ssm_meta,
    iterations = 5,
    showprogress=true,
)

xmarginals  = result.posteriors[:x][1] # why is this a 1-element vector?

let 
    fig = Figure(size=(800,400))
    ax1 = Axis(fig[1,1])
    lines!(t_sim, getindex.(x, 1), label = "Hidden Signal (dim-1)", color = :orange)
    lines!(t_sim, getindex.(x, 2), label = "Hidden Signal (dim-2)", color = :green)
    band!(t_sim, getindex.(mean.(xmarginals), 1) .- (getindex.(var.(xmarginals), 1) .|> sqrt), getindex.(mean.(xmarginals), 1) .+ (getindex.(var.(xmarginals), 1) .|> sqrt), label = "Estimated Signal (dim-1)", color = (:violet, 0.5))
    band!(t_sim, getindex.(mean.(xmarginals), 2) .- (getindex.(var.(xmarginals), 2) .|> sqrt), getindex.(mean.(xmarginals), 2) .+ (getindex.(var.(xmarginals), 2) .|> sqrt), label = "Estimated Signal (dim-2)", color = (:teal, 0.5))
    lines!(t_sim, getindex.(mean.(xmarginals), 1), label = "Estimated Signal (dim-1)", color = :violet)
    lines!(t_sim, getindex.(mean.(xmarginals), 2), label = "Estimated Signal (dim-2)", color = :teal)
    Legend(fig[1,2], ax1, merge=true, unique=false)
    fig
end
``` 
![fig](https://github.com/ReactiveBayes/RxInfer.jl/assets/50108075/c0a92ab2-3113-4009-8019-c432ce1df07b)

**Reply by ValentinKaisermayer** [reply]
If I crank up the `iterations` to 50 it works. 

![fig](https://github.com/ReactiveBayes/RxInfer.jl/assets/50108075/0a0bdc58-1154-4092-82af-362df09af0b2)

```julia
let
    fig = Figure(; size=(400, 300))
    ax1 = Axis(fig[1, 1])
    lines!(result.free_energy)
    save("fig.png", fig)
    fig
end
```

![fig](https://github.com/ReactiveBayes/RxInfer.jl/assets/50108075/dc5ea6fd-4516-4c10-b678-3952bca005b7)

However, it is painfully slow. A KF implementation could do this in milliseconds. Is there a way to improve this?
How could I infer `p.m`?

**Reply by albertpod** [reply]
Hi @ValentinKaisermayer. Well, the thing is, you ain't doing KF here but rather KF smoother. You feed the whole signal into `infer` function, I suggest checking out the following [example](https://reactivebayes.github.io/RxInfer.jl/stable/examples/advanced_examples/Infinite%20Data%20Stream/).
Another possible source of painful slowness is `UT()`, it's not very well optimized.

**Reply by ValentinKaisermayer** [reply]
Thanks for your help. 
My actual goal would be parameter inference, so I want to consider at least large portions of the data at once. Is `CTransition` of any help here?

**Reply by albertpod** [reply]
Yes, makes sense. `ContinuousTransition` node can indeed be helpful as long as you can frame your transition as $$x_t=Ax_{t-1}$$.

By the way, how does Linearization perform in your case?
Another question: doesn't the inference run without an initialization parameter in your case?

**Reply by bvdmitri** [reply]
Thanks for your question!

It's not entirely fair to compare with the Kalman Filter since your current approach is smoothing on a full graph with 12,000 observations, as outlined by @albertpod. Each iteration requires a full forward and backward pass, involving many approximations. Specifically, it needs to call the multivariate Unscented transform 48,000 times per iteration.

If you want to adapt this to the filtering case (using only the forward pass), I recommend this [material](https://reactivebayes.github.io/RxInfer.jl/stable/manuals/inference/streamlined/) in addition to the one suggested by Albert. Filtering should indeed improve performance.

Additionally, this [hidden example](https://github.com/ReactiveBayes/RxInfer.jl/blob/main/examples/hidden_examples/Tiny%20Benchmark.ipynb) demonstrates the significant performance difference between filtering and smoothing. For a small dataset of univariate (!) 500 observations with no approximations, filtering shows a 30x performance improvement. Let us know if you need further assistance in adapting your example to the filtering case.

Another suggestion: you might use the `dot` function with a standard basis vector as the observation function. This approach can eliminate the need for the `g` function and the Unscented transform, improving accuracy and convergence speed, and removing the need for initializing messages and performing iterations. We have specialized implementations for dot that compute messages faster and more accurately, for example:

```julia
@model function ssm(y, f, g, tS, x0, u, t, Q, R, p)
    x_prior ~ x0
    c = ReactiveMP.StandardBasisVector(2, 1)
    x_prev = x_prior
    for i in eachindex(y)
        x[i] ~ MvNormal(μ=f(x_prev, u[:,i], p, t[i], tS), Σ=Q)
        y[i] ~ Normal(μ=dot(c, x[i]), var=R[1, 1])
        x_prev = x[i]
    end
end

prior_x = MvNormalMeanCovariance(x[1], diageye(2))

model = ssm(f=f, g=g, tS=tS, x0=prior_x, u=u_sim, t=t_sim, Q=Q, R=R, p = p)

ssm_meta = @meta begin 
    f() -> Unscented()
end
```

Additionally, avoid using the global variable `p`; instead, pass it as an argument to the model function (and any other variables I may have missed). Otherwise, Julia gets confused and checks its type every time it is accessed, which significantly slows down performance.

Also, consider checking `Linearization` instead of `Unscented` for potentially faster performance. (beware global variables, it makes it even slower when using `Linearization` because it utilizes `ForwardDiff`, which performs poorly with global variable access.)

By replacing `g` with dot and making `p` local to the model function, I was able to reduce inference times on my computer from 150 seconds per iteration to about 50 seconds. Not that bad if you consider full forward and full backward pass with 12,000 observations.
```julia
@time result = infer(
    model = model,
    data  = (y = first.(y), ), 
    options = (limit_stack_depth = 200, ), 
    # initialization = imessages,
    meta = ssm_meta,
    iterations = 1,
    showprogress = true,
)
```

The inference also required only one iteration and converged to the following:

![Screenshot 2024-05-17 at 18 27 55](https://github.com/ReactiveBayes/RxInfer.jl/assets/6557701/2f77dd48-f1aa-40aa-894a-630184b6eafb)

> `result.posteriors[:x][1] # why is this a 1-element vector?`

it depends on the number of iterations. Perhaps you ran it with a single iteration?

> options = (limit_stack_depth = 10, ), 

This setting is very low and can significantly slow down inference. It's recommended to set it as high as your computer allows; I typically use 500-600. However, in this example, I found 250 to be sufficient. This setting is usually not required for the filtering case.

> How could I infer p.m?

To infer `m`, you can treat it as a standalone random variable, assign a prior to it, and infer it like any other variable. We do not support inference over components of a named tuple, so it should be a separate variable.

Please let us know if you have other question. I or my colleagues will be happy to help you.

**Reply by bvdmitri** [reply]
I would also emphasize that while filtering is generally faster, it is often less accurate. The difference in accuracy should be evaluated on a case-by-case basis.

Additionally, as @albertpod mentioned, our current Unscented approximation technique requires some refactoring and is quite slow. We would be happy to assist if you choose RxInfer and decide to work on improving it. We are willing to guide you through the process.

**Reply by Nimrais** [reply, marked as answer]
@ValentinKaisermayer I decided to play with your model a bit to also infer the p parameter and employ @bvdmitri's suggestion to remove the g function. For me, the inference now runs for 20 seconds, taking 1 second per iteration.

```julia
using RxInfer
using Random
using LinearAlgebra
using CairoMakie

function f(x, u, p, t, tS)
    return x + tS*[
        x[2]
        -p[1] / p[3] * x[1] - p[2] / p[3] * x[2] - 1 / p[3] * u[1]
    ]
end

g(x, u, p, t) = [x[1]]

tS = 1/100
t_sim = 0:tS:120
u_sim = reduce(
    hcat,
    map(t -> [exp(-0.1t) * sin(1 / 10 * 2 * pi * t) + sin(1 / 20 * 2 * pi * t)], t_sim),
)
x0 = [0, 0]

function generate_data(rng, f, g, tS, x0, u, t, Q, R, p)
    x_prev = x0

    x = Vector{Vector{Float64}}(undef, length(t))
    y = Vector{Vector{Float64}}(undef, length(t))

    for i in 1:length(t)
        x[i] = rand(rng, MvNormal(f(x_prev, u[:,i], p, t[i], tS), Q))
        y[i] = rand(rng, MvNormal(g(x[i], u[:,i], p, t[i]), R))
        x_prev = x[i]
    end

    y = map(yi -> yi[1], y)
    
    return x, y
end

seed = 1234

Q = 0.1 .* diageye(2)
R = 10 .* diageye(1)
p_vec = [1, 0.1, 0.8]

x, y = generate_data(MersenneTwister(1234), f, g, tS, x0, u_sim, t_sim, Q, R, p_vec)

let 
    fig = Figure()
    ax1 = Axis(fig[1,1])
    lines!(t_sim, getindex.(x, 1), label = "Hidden Signal (dim-1)", color = :teal)
    lines!(t_sim, getindex.(x, 2), label = "Hidden Signal (dim-2)", color = :violet)
    ax2 = Axis(fig[2,1])
    lines!(t_sim, getindex.(y, 1), label = "Observation", color = :orange)
    fig
end

@model function ssm(y, f, tS, x0, u, t, Q, R, prior_m)
    x_prior ~ x0
    c = ReactiveMP.StandardBasisVector(2, 1)
    x_prev = x_prior
    m ~ prior_m
    for i in eachindex(y)
        x[i] ~ MvNormal(μ=f(x_prev, u[:,i], m, t[i], tS), Σ=Q)
        y[i] ~ Normal(μ=dot(c, x[i]), var=R[1, 1])
        x_prev = x[i]
    end
end

prior_x = MvNormalMeanCovariance(x[1], diageye(2))
prior_m = MvNormalMeanCovariance([0.95, 0.09, 0.7], diageye(3)*0.01)

model = ssm(f=f, tS=tS, x0=prior_x, u=u_sim, t=t_sim, Q=Q, R=R, prior_m = prior_m)

ssm_meta = @meta begin 
    f() -> Unscented()
end

imessages = @initialization begin
    μ(m) = prior_m
end

result = infer(
    model = model,
    data  = (y=y, ), 
    options = (limit_stack_depth = 250, ), 
    meta = ssm_meta,
    initialization = imessages,
    iterations = 20,
    showprogress=true,
    free_energy=true
)

lines(1:20, result.free_energy, label="Free Energy")

xmarginals  = result.posteriors[:x][20]

let 
    fig = Figure(size=(800,400))
    ax1 = Axis(fig[1,1])
    lines!(t_sim, getindex.(x, 1), label = "Hidden Signal (dim-1)", color = :orange)
    lines!(t_sim, getindex.(x, 2), label = "Hidden Signal (dim-2)", color = :green)
    band!(t_sim, getindex.(mean.(xmarginals), 1) .- (getindex.(var.(xmarginals), 1) .|> sqrt), getindex.(mean.(xmarginals), 1) .+ (getindex.(var.(xmarginals), 1) .|> sqrt), label = "Estimated Signal (dim-1)", color = (:violet, 0.5))
    band!(t_sim, getindex.(mean.(xmarginals), 2) .- (getindex.(var.(xmarginals), 2) .|> sqrt), getindex.(mean.(xmarginals), 2) .+ (getindex.(var.(xmarginals), 2) .|> sqrt), label = "Estimated Signal (dim-2)", color = (:teal, 0.5))
    lines!(t_sim, getindex.(mean.(xmarginals), 1), label = "Estimated Signal (dim-1)", color = :violet)
    lines!(t_sim, getindex.(mean.(xmarginals), 2), label = "Estimated Signal (dim-2)", color = :teal)
    Legend(fig[1,2], ax1, merge=true, unique=false)
    fig
end
```

The state estimation seems okay, however it's quite off now for the second dimension of the hidden signal. I think it's due my `prior_m` choice.
You can try to play with it.

<img width="462" alt="Screenshot 2024-05-17 at 20 09 01" src="https://github.com/ReactiveBayes/RxInfer.jl/assets/29106356/29f9c4d3-0b66-47b3-9a12-4a1314514e0c">

The inferred mean:

julia> mean(result.posteriors[:m][20])
3-element Vector{Float64}:
 2.8187083107291464
 1.9576793500593177
 1.5808318658894194

**Reply by bvdmitri** [reply]
Great, many thanks @Nimrais ! How long does it take for you if you don't compute the free energy? It's generally not necessary for the inference, only for checking if the algorithm has converged or not.

**Reply by bvdmitri** [reply]
I would also assume that the `R` (which is the observational noise) has quite a big value. It was the same in the original formulation from @ValentinKaisermayer, but I can see that the scale of the data is somewhat `(-20, 20)` and the observational noise is set to `10`. That might also explain the difference in the inferred mean. Does it infer better if you decrease `R` from `10` to let's say `1` or even `0.1`?

**Reply by Nimrais** [reply]
@bvdmitri Yes, by reducing the `R` value, I can obtain improvement for the state estimation part, which makes sense as the noise is quite strong in the original formulation.

Here, I reduced the `R` to 1

![Screenshot 2024-05-17 at 21 03 57](https://github.com/ReactiveBayes/RxInfer.jl/assets/29106356/76248b11-7cec-454d-910f-894c0061d9ea)

However, the free energy does not look smooth. I think it's the approximation happening around the `m` edge that does not really perform well.

![Screenshot 2024-05-17 at 21 18 24](https://github.com/ReactiveBayes/RxInfer.jl/assets/29106356/7120aea2-535b-473e-b3fb-35b252843c01)

Regarding the free energy computation overhead, it does not seem to really contribute to the computation time: around 1.13 sec per iteration with the FE computation, and around 0.9 sec per iteration without.

**Reply by Nimrais** [reply]
@albertpod In the model with the inference for p using the Linearization() meta, the inference is diverging.

![Screenshot 2024-05-17 at 21 32 14](https://github.com/ReactiveBayes/RxInfer.jl/assets/29106356/473c387d-bf44-43b3-8dfe-41f5c2fc2dc1)

**Reply by ValentinKaisermayer** [reply]
@Nimrais Very cool! 😎 

Why do you use a MvNormal prior for the parameters and not individual distributions? I would guess that it is reasonable to assume that they are uncorrelated. Could you add a constraint for that?
Also, all parameters need to be strictly positive, otherwise the ODE system is unstable. In this [post](https://github.com/orgs/ReactiveBayes/discussions/300) a log transformation is used for that, but aren't there truncated distributions for that?

**Reply by bvdmitri** [reply]
>Why do you use a MvNormal prior for the parameters and not individual distributions? I would guess that it is reasonable to assume that they are uncorrelated. Could you add a constraint for that?

You could create three different variables, say `p1`, `p2`, and `p3`, and assign a univariate normal prior distribution to each. The function would also need to be modified to accept each variable individually.

> Also, all parameters need to be strictly positive, otherwise the ODE system is unstable. In this [post](https://github.com/orgs/ReactiveBayes/discussions/300) a log transformation is used for that, but aren't there truncated distributions for that?

This is more challenging, as we don't have out-of-the-box support for truncated distributions. Another solution (in addition to the post) could be to use the CVI approximation along with the Gamma distribution as a prior. We are planning a major overhaul of the CVI approximations, as the current version doesn't always yield stable results.

**Reply by albertpod** [reply]
> This is more challenging, as we don't have out-of-the-box support for truncated distributions. Another solution (in addition to the post) could be to use the CVI approximation along with the Gamma distribution as a prior. We are planning a major overhaul of the CVI approximations, as the current version doesn't always yield stable results.

The hack I sometimes use to prioritize speed is to pass a Gaussian through a delta function that truncates the variable and bypasses it to the function where positivity is required.

**Reply by bvdmitri** [reply]
> The hack I sometimes use to prioritize speed is to pass a Gaussian through a delta function that truncates the variable and bypasses it to the function where positivity is required.

Or exponentiate (or `clamp`) the input inside the `f` function, by doing that the value inside the `f` function will always be positive.

**Reply by ValentinKaisermayer** [reply]
[Turing](https://turing.ml/dev/docs/using-turing/advanced#domain-transformation) uses bijectors for this. 
So RxInfer.jl does not use Distributions.jl?

**Reply by bvdmitri** [reply]
@ValentinKaisermayer no, RxInfer is a fundamentally different approach to Bayesian inference. Turing does sampling while RxInfer does message passing. There are both pros and cons for both approaches. Currently there is no easy way to integrate bijectors with the message passing technique, but we have some cool on-going research that will expand the class of supported models in the near future. Stay tuned or join [our monthly meetings](https://dynalist.io/d/F4aA-Z2c8X-M1iWTn9hY_ndN) if you want to hear the progress and/or ask more questions :)

**Reply by bvdmitri** [reply]
@ValentinKaisermayer This [material](https://reactivebayes.github.io/RxInfer.jl/stable/manuals/comparison/#comparison) might be interesting for you to read how RxInfer is different from other solutions

**Comment by ValentinKaisermayer** [comment]
I tested the solution, and the progress bar shows 8 s, however, `@time` of `ìnfer`  showed 223 s.

**Reply by bvdmitri** [reply]
Hm, interesting. Do you measure the compilation time? `@time` should report these, and while compilation can take a significant portion of the runtime, it should typically occur only once. However, it might happen every time if Julia is confused for some reason. If `@time` doesn't show excessive compilation times, the issue might be with the model creation, although it shouldn't take that long. You could use [callbacks](https://reactivebayes.github.io/RxInfer.jl/stable/manuals/inference/static/#manual-static-inference-callbacks) to measure the exact time of each step in the procedure. For example you can use `before_model_creation` and `after_model_creation` callbacks to measure time between those. This would help us understand why it's taking so much time, even though the progress bar only reports 8 seconds.
